### ISSUE TITLE
<table> moved out of <p> tag

### ISSUE THREAD

**Issue by bapilars** [htacg/tidy-html5#969]
Hello Team,

i have a problem regarding block level tags inside a paragraph.

Depending on the doctype the browsers show the `<table>` inside the `<p>`, or close the `<p>` before the `<table>`.
- `<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">`: table allowed inside p
- `<!DOCTYPE html>`: table not allowed inside p

Input file:
```
<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<title></title>
</head>
<body>
<p style="background-color:green;margin:50px;padding:50px">
    <table>
        <tr>
            <td style=
                "background:#000099;color:#f3f3f3;font-family:Helvetica,Arial,sans-serif;padding:10px">
                Hello World</td>
        </tr>
    </table>
</p>
</body>
</html>
```

As far as i have seen there is no way to configure tidy to behave in the same way as the browser for this case.

I saw in the code that there is an `TY_(AdjustTags)` function to change behaviour depending on HTML4 or HTML5, if i extend it by
```
np = (Dict *)TY_(LookupTagDef)( TidyTag_P );
if (np)
{
    np->parser = TY_(ParseBlock);
}
```
this issue will be solved, but i am not sure whether it is entirely correct because `<p>` seems to be some kind of special block (optional end tag?) that should close automatically when another `<p>` is started. 

The code change would keep such structure unchanged:
```
<p style="color:red">
    new paragraph, red - not closed
    <p style="color:blue">another paragraph, blue - closed</p>
    continue red paragraph, close it
</p>
```
but the first red paragraph should probably close before the second blue one (as tidy does without my change), and the browser also does it like this and shows last text in black.

So my question to you is whether you had maybe already encountered a similar issue, or whether this is really an issue that can be fixed (to allow the table in p, but still close p before a next starting p), and generally what is the correct behaviour in your eyes.

**Comment by balthisar** [comment]
@bapilars, thanks for the question, and the sample code. If you want to test your code in more depth, what happens when you run the regression tests? What we don't want to do is fail a bunch of test cases without really good reason.

I'm guessing that the block level tags are allowed because HTML 4 transitional is kind of "HTML 4 sloppy". 

You're asking about the behavior in HTML5, though, and to me "correct (default) behavior" means do whatever the specification requires. If I run your sample in the [W3C Nu Validator](https://html5.validator.nu/) with `<!DOCTYPE html>`, it complains "Error: No p element in scope but a p end tag seen." That suggests to me that it's assuming that the `<table>` is not nested in the `<p>` but is actually _closing_ the `<p>`, because it's special and doesn't need to be closed.

Finally, if I open your file in Chrome and look at the actual DOM structure, I see this:

![image](https://user-images.githubusercontent.com/3053246/125622000-965a0668-885b-4b1c-872c-595f0d706e78.png)

So the browser is doing the same thing as Tidy and the validator are: it's closing the `<p>`, then following it with the `<table>`, and fixing the original `<\p>` by changing it to `<p></p>`.

My opinion is that Tidy is doing the right thing for HTML5. If you have something to change my mind, let me know, but if you agree, also let me know so we can close this.

**Comment by bapilars** [comment]
@balthisar thanks for your fast reply. I noticed also this browser behaviour with the `<DOCTYPE html>`, and i agree it is correct for HTML5. But the visual difference in my testfile is really huge: 

HTML4 doctype:
![html4_doctype](https://user-images.githubusercontent.com/87423369/125623892-199f72eb-cfbd-4bea-a749-ce91e09e311f.png)

vs HTML 5:
![html_doctype](https://user-images.githubusercontent.com/87423369/125623934-c3897679-2d66-40b6-8d8b-75ac6af74729.png)

For our use case we would like to minimize the visual difference from input to output, but still get a kind of xhtml output (so a well formed XML structure) that we can process later on. And thanks to this great library it works really good - i know this is a pretty special case, i did not run regression tests but i am sure there will be a lot so this change without further conditional checks is not acceptable (as in my second example, where the first p should close but doesn't anymore with the change).

Is the tidy output format always in HTML5 and can not (..is not indented to) be configured to use the same doctype for the xhtml output?

**Comment by balthisar** [comment]
@bapilars, What's the doctype that you want to use? When I use tidy with transitional doctype, you're right, it's not acting the same way as the browser. If you're suggesting that Tidy allow table nest within paragraphs for transitional doctype, then I think I agree.

**Comment by bapilars** [comment]
@balthisar Yes exactly, i suggest that tidy could be maybe extended somehow to support changes in the nesting hierarchy depending on the input doctype, or some other configuration. I didn't check in detail what other special things there might be except the `<table>`. 

My background is, we do a xhtml formatting to output for example pdf files, and can usually not really influence the source or generation of the html files that we get as input, so our goal is independent from any doctype to get ideally the same visual result in the output pdf as the browser shows when opening the input html.

**Comment by balthisar** [comment]
@bapilars, have you tried the `input-xml` option? If I set it to yes, even with an HTML5 doctype, I get the following output:

~~~
<!DOCTYPE html>
<html>
<head>
<title></title>
</head>
<body>
<p style="background-color:green;margin:50px;padding:50px">
<table>
<tr>
<td style="background:#000099;color:#f3f3f3;font-family:Helvetica,Arial,sans-serif;padding:10px">
Hello World</td>
</tr>
</table>
</p>
</body>
</html>
~~~

I think this is exactly what you're looking for?

**Comment by bapilars** [comment]
@balthisar yes i tried it, but then tidy doesn't clean up the file anymore, or? In this example it works because it is well formed XML, but for example if tidy should repair a file due to a missing closing tag, illegal nesting, or a `<br>` it will not work anymore with a message like `unexpected </body> in <br>`.

**Comment by balthisar** [comment]
Understood. I'm open to a proposal wherein Tidy behaves like a browser for HTML 4 transitional, but I need a proposal. I can't check every iteration of HTML4 transitional, especially as the "transition" should already be over and done with many years ago ;-)

**Comment by bapilars** [comment]
As long as the browser engines still support it the "transition" will probably not be over ;). Do you mean an implementation design proposal?

**Comment by balthisar** [comment]
An actual PR would be great, otherwise a design proposal, but really, even just a complete functional description of what you'd like to happen. For example, you've said:

> I didn't check in detail what other special things there might be except the `<table>`.

If your proposal is table, then you're already close to a PR. If you're looking for other things in addition to table, I can only say that I won't be the one hunting them down.

**Comment by bapilars** [comment]
Okay fine i will see what i can do, i didn't expect you to hunt it down, i just wanted to keep it more general and not restricted to only table in case other issues related to the transitional doctype appear in the future. Thank you very much so far :)

**Comment by balthisar** [comment]
> i just wanted to keep it more general and not restricted to only table in case other issues related to the transitional doctype appear in the future

Thanks. We try to straddle the balance between not affecting browser rendering (which your proposal for table fixes), and keeping to the standards. By "chasing things down" I mean that you've sold me table, but there's no way I can check every single block level tag to see how different browsers behave for different versions of HTML. Hope that clarifies my meaning!

**Comment by AnrDaemon** [comment]
Even transitional doctype disallows block-level tags inside paragraphs.
You can try declaring `<table>` as inline tag, though.
